examples/main/main.cpp

```diff
@@ -158,7 +158,9 @@ int main(int argc, char ** argv) {
     }
 
     const llama_vocab * vocab = llama_model_get_vocab(model);
-    auto chat_templates = common_chat_templates_from_model(model, params.chat_template);
+    common_chat_templates_ptr chat_templates(
+        common_chat_templates_init(model, params.chat_template),
+        &common_chat_templates_free);
```

Comment:
This should also work:

```suggestion
    common_chat_templates_ptr chat_templates(common_chat_templates_init(model, params.chat_template));
```

Reply:
This unique_ptr is templated w/ the free function's type and its ctor does need the function pointer.

```c++
typedef std::unique_ptr<struct common_chat_templates, decltype(&common_chat_templates_free)> common_chat_templates_ptr;
```

(a raw `std::unique_ptr<struct common_chat_templates>` wouldn't work as the caller won't know how to delete the opaque struct; possible workaround would be to use a base class w/ virtual dtor but then quite cumbersome)

Now returning it from `common_chat_templates_init` directly (https://github.com/ggml-org/llama.cpp/pull/11900/commits/da0982a00e0011bcd6a03b22cdf47f33992151fd), which simplifies usage.

Reply:
Probably we can update the `typedef` similar to what we have in `ggml-cpp.h`:

https://github.com/ggml-org/llama.cpp/blob/master/ggml/include/ggml-cpp.h#L15-L22

This would avoid having to explicitly specify the deleter?

Reply:
Neat, thanks, updated.